arwen/cryptoapi/cryptoei.go

```diff
@@ -0,0 +1,266 @@
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.SHA256
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, length)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	result, err := crypto.Sha256(data)
+	if err != nil {
+		return 1
+	}
+
+	err = runtime.MemStore(resultOffset, result)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	return 0
+}
+
+//export keccak256
+func keccak256(context unsafe.Pointer, dataOffset int32, length int32, resultOffset int32) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.SHA256
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, length)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	result, err := crypto.Keccak256(data)
+	if err != nil {
+		return 1
+	}
+
+	err = runtime.MemStore(resultOffset, result)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	return 0
+}
+
+//export ripemd160
+func ripemd160(context unsafe.Pointer, dataOffset int32, length int32, resultOffset int32) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.Ripemd160
+	metering.UseGas(gasToUse)
+
+	data, err := runtime.MemLoad(dataOffset, length)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	result, err := crypto.Ripemd160(data)
+	if err != nil {
+		return 1
+	}
+
+	err = runtime.MemStore(resultOffset, result)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	return 0
+}
+
+//export verifyBLS
+func verifyBLS(
+	context unsafe.Pointer,
+	keyOffset int32,
+	messageOffset int32,
+	messageLength int32,
+	sigOffset int32,
+) int32 {
+	runtime := arwen.GetRuntimeContext(context)
+	crypto := arwen.GetCryptoContext(context)
+	metering := arwen.GetMeteringContext(context)
+
+	gasToUse := metering.GasSchedule().CryptoAPICost.VerifyBLS
+	metering.UseGas(gasToUse)
+
+	key, err := runtime.MemLoad(keyOffset, blsPublicKeyLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	message, err := runtime.MemLoad(messageOffset, messageLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	sig, err := runtime.MemLoad(sigOffset, blsSignatureLength)
+	if arwen.WithFault(err, context, runtime.CryptoAPIErrorShouldFailExecution()) {
+		return 1
+	}
+
+	invalidSigErr := crypto.VerifyBLS(key, message, sig)
+	if invalidSigErr != nil {
+		return 1
```

Comment:
Our API returns `1` for internal errors at the moment (or positive integers). By contrast, `invalidSigErr` means "signature validation was executed successfuly, with the answer «invalid»". Maybe return `-1` here instead, so a smart contract could distinguish between a failure and an invalid signature. Optional.

Reply:
You are right, it's better to distinguish from internal errors vs `invalid sig`. Done